### PR TITLE
REF use GSL gamma function

### LINKED ISSUE #570: potential numerical issues in the fftlog implementation
Right now to compute the log of the gamma function, we compute the gamma function and then take the log of it. This procedure is not great numerically IIUIC. 

I think we should either switch to using the GSL gamma functions or reimplement so that we compute the log first and then exponentiate if needed.

### PR DESCRIPTION
**Pull request by beckermr**
This PR removes our home grown gamma function in favor of that from GSL.

Closes #570

### CODE REVIEW

File: tests/ccl_test_correlation_3d.c [hunk 1]
```diff
@@ -82,16 +82,17 @@ static void compare_correlation_3d(int i_model,struct corrs_3d_data * data)
   }
   nr=linecount(f)-1; rewind(f);
 
+  // FIXME: these are not real standards
   // tolerence on abs difference in r^2 xi(r) for the range r = 0.1 - 100 Mpc (40 points in r) for z=0,1,2,3,4,5
-  double CORR_TOLERANCE1[6] = {2.84e-2, 3.57e-3, 1.37e-3, 1.26e-3, 2.0e-4, 1.8e-4};
+  double CORR_TOLERANCE1[6] = {2.84e-2, 3.6e-3, 1.37e-3, 1.26e-3, 2.5e-4, 1.8e-4};
```
Comment (philbull) [inline tests/ccl_test_correlation_3d.c]: So has there been a precision regression?

File: tests/ccl_test_correlation_3dRSD.c [hunk 2]
```diff
@@ -83,16 +83,17 @@ static void compare_correlation_3dRSD(int i_model,struct corrs_3dRSD_data * data
   }
   nr=linecount(f)-1; rewind(f);
 
+  // FIXME: these are not real standards
   // tolerence on abs difference in r^2 xi(r) for the range r = 0.1 - 100 Mpc (40 points in r) for z=0
-  double CORR_TOLERANCE1 = 0.06;
+  double CORR_TOLERANCE1 = 0.08;
```
Comment (philbull) [inline tests/ccl_test_correlation_3dRSD.c]: Another change to validated tolerance. Needs to be double-checked.

**Comment by beckermr** [comment]
The change in tolerances with respect to the benchmarks does not mean we have lost precision. My guess is that CCL is now better than the benchmarks. IDK how the benchmarks were validated but they are numerical computations just like CCL and can suffer from the same issue.

**Comment by philbull** [comment]
That's fine, but we still need to check if this changes anything significant (e.g. with reference to the CCL paper). We shouldn't change the benchmark tolerances just to get the unit tests to pass.

**Comment by beckermr** [comment]
It is very clear the benchmark tolerances were set to make the tests pass in the first place. Otherwise why would the code have values like 3.57e-3?

**Comment by beckermr** [comment]
Here is a simple comparison of the implementations. The GSL one is definitely better.

```
(anl) localhost:gsl_tests Matt$ ./test_lngamma 
val:    -1.00000000000000000000000000000000000000000000000000e+01 +      1.00000000000000000000000000000000000000000000000000e+00i
    ours:    -1.63587472093715184939810569630935788154602050781250e+01 +     7.82460502739822816309356312558520585298538208007812e-01i
    gsl :    -1.63587472093715149412673781625926494598388671875000e+01 +     7.82460502739823038353961237589828670024871826171875e-01i
val:    1.00000000000000002505909183520875968569614680770371e-300 +      0.00000000000000000000000000000000000000000000000000e+00i
    ours:     6.90775527898213681510242167860269546508789062500000e+02 +     0.00000000000000000000000000000000000000000000000000e+00i
    gsl :     6.90775527898213681510242167860269546508789062500000e+02 +    -0.00000000000000000000000000000000000000000000000000e+00i
val:     1.00000000000000000000000000000000000000000000000000e+01 +      1.00000000000000000000000000000000000000000000000000e+00i
    ours:     1.27493405523471707851967948954552412033081054687500e+01 +     2.25358817704355685052064472984056919813156127929688e+00i
    gsl :     1.27493405523471707851967948954552412033081054687500e+01 +     2.25358817704355685052064472984056919813156127929688e+00i
val:     1.00000000000000000000000000000000000000000000000000e+02 +      0.00000000000000000000000000000000000000000000000000e+00i
    ours:     3.59134205369575340682786190882325172424316406250000e+02 +     0.00000000000000000000000000000000000000000000000000e+00i
    gsl :     3.59134205369575340682786190882325172424316406250000e+02 +     0.00000000000000000000000000000000000000000000000000e+00i
val:     1.00000000000000000000000000000000000000000000000000e+03 +      0.00000000000000000000000000000000000000000000000000e+00i
    ours:                                                          nan +                                                          nani
    gsl :     5.90522042320918080804403871297836303710937500000000e+03 +     0.00000000000000000000000000000000000000000000000000e+00i
```

Code is [here](https://gist.github.com/beckermr/0f9b4a95ecd4e1f0ac3cd66767f3526e).

**Comment by philbull** [comment]
Just want to make sure the relevant people are aware that this is being changed (there were quite lengthy previous discussions about choosing non-ideal tolerance values in certain cases).

**Comment by beckermr** [comment]
@zdu863 You should look through this and comment if you'd like.

**Comment by beckermr** [comment]
Bump here @philbull

**Comment by beckermr** [comment]
Bump again here @philbull @zdu863

**Comment by beckermr** [comment]
Bump again here @philbull @zdu863

**Comment by beckermr** [comment]
We've stalled on this PR for a month. Can someone please approve this?

Bump  @philbull @zdu863

**Comment by beckermr** [comment]
Bump again  @philbull @zdu863

**Pull request merged**